### ISSUE TITLE
Tool call JSON malformation under MQ4 quantization — Qwen3.6 27B

### ISSUE THREAD

**Issue by Fluorax** [Kaden-Schutt/hipfire#111]
# Tool call JSON malformation under MQ4 quantization — Qwen3.6 27B

## Environment

- **GPU:** RX 7900 XTX (gfx1100, 24GB), ROCm 6.2
- **hipfire:** built from source, master c448d5e+
- **Model:** `qwen3.6-27b.mq4`
- **Client:** pi agentic harness (OpenAI-compatible tool calling)

## Problem

The model consistently produces malformed tool call syntax inside `<tool_call>` blocks. Instead of valid JSON:

```
<tool_call>
{"name": "write", "arguments": {"path": "/home/mike/rate_limiter.py", "content": "..."}}
</tool_call>
```

It produces XML-like mixed syntax:

```
<tool_call>
<plain>write</param> {"param": "/home/mike/rate_limiter.py", "content": "..."}
</tool_call>
```

The `parseToolCalls` function in `cli/index.ts` cannot parse this — it expects valid JSON inside the tags. The tool call silently fails, the model believes it succeeded, and the file is never written.

## Reproduction

1. Connect hipfire to an OpenAI-compatible agentic harness with file write tools
2. Ask the model to implement and save a Python class
3. Observe `<tool_call>` block with malformed JSON in the response
4. Tool call fails to execute despite model reporting success

## Comparison

Same prompt on Qwen3.6 27B GGUF via llamacpp produces correctly formatted tool calls:

```
<tool_call>
{"name": "write", "arguments": {"path": "/home/dev/rate_limiter.py", "content": "..."}}
</tool_call>
```

File is written successfully. This confirms the base model weights know the correct format, the malformation is specific to MQ4.

## Hypothesis

MQ4 FWHT rotation shifts the token probability distribution enough to corrupt structured output at the token boundaries of JSON syntax (`{`, `"`, `:`, `}`). The model produces plausible-looking but invalid structured tokens that visually resemble the correct format.

This may be related to the same mechanism causing repetition attractors on long structured output, both are failures of precise token prediction under quantization pressure.

## Expected behaviour

Tool calls should produce valid JSON inside `<tool_call>` tags, parseable by `parseToolCalls` in `cli/index.ts`.

## Potential fix paths

1. **MQ6**  higher quality quantization, less distribution shift on structured tokens
2. **Post-processing**  attempt to repair malformed JSON in `parseToolCalls` before discarding
3. **Tool call fine-tuning calibration**  include tool call examples in the MQ4 calibration dataset to preserve structured output precision

**Comment by Kaden-Schutt** [comment]
Reproduced on master locally on a 7900 XTX (gfx1100) using `qwen3.6:27b.mq4` at temp=0 with the OpenAI v1/chat/completions streaming endpoint and a 4-tool harness shape (write/read/bash/edit + system prompt). The malformation surface in my capture is slightly different from yours but the same class:

- Your output: `<plain>write</param> {ARGS}` (XML-tag-shape corruption inside the `<tool_call>` block).
- My capture: `{"name": "write", "path": "...", "content": "..."}` (flat JSON; sibling args, no `arguments` wrapper).

Both pass through the existing `parseToolCalls`, which then does `JSON.stringify(tc.arguments || {})` and writes `"{}"`, silently dropping the call payload. The downstream harness gets a function call with empty args and the file is never written.

**Root cause:** MQ4 weight quantization drift on structured tokens. The argmax margin for `{` / `"` / `:` / tag-name fragments is narrow, so the FWHT-rotated 4-bit weights flip greedy-decode argmax there before they flip narrative tokens. Same shape as #87 (Q8_1 activation drift on Wo row 3994), different layer in the stack. Single-tool / low-context prompts emit clean nested JSON; the malformation emerges with multi-tool + system prompt context (which shifts the per-position activation distribution).

**Tonight's fix (shipped to master `62e5767`, branch `fix/111-tool-call-mq4-malformation`):** defensive `parseToolCalls` upgrade that handles three shapes:

1. Spec-compliant `{"name": ..., "arguments": {...}}` (passes through).
2. Flat `{"name": ..., field1: ..., field2: ...}` (coerce siblings into `arguments`, dropping known-metadata keys: type, id, function).
3. XML-tag corruption: `<plain>NAME</param>`, `<function=NAME>`, or `<tool name=NAME>` followed by a JSON object. Extract the name from the tag and the args via balanced-brace JSON scan (handles braces inside JSON strings, escaped quotes).

When any block is repaired, `parseToolCalls` emits `[hipfire] tool_call: repaired N malformed block(s) (MQ4 #111 stopgap)` to stderr so harness logs flag the recovery.

Verification:
- 10 unit tests in `cli/parse_tool_calls.test.ts` covering control + flat + XML + escaped quotes + unparseable args fallback.
- Re-ran the multi-tool streaming repro: `finish_reason` flipped from `"stop"` (silent drop) to `"tool_calls"`, and the tool_calls delta now carries the actual file payload as `function.arguments`.

**This is a stopgap; the proper fix is calibration.** Tracked in `MANUAL_REVIEW.md`: requantize MQ4 with tool-call-shape calibration samples (OpenAI tool messages + ChatML `<tool_call>` blocks) weighted ~5x. If that cleans the structured-token tail without regressing wikitext PPL, ship a `mq4` rev bump and reupload the family. If it doesn't, escalate to MQ4-Lloyd (Lloyd-Max codebook variant of MQ4).

Leaving this issue open for the calibration root-cause.

Pull master and your existing `qwen3.6:27b.mq4` artifact should now produce structurally valid `tool_calls` deltas with non-empty arguments. If you continue to see specific malformation patterns my repair doesn't catch, please paste the raw `<tool_call>` block (the inner text between the tags) and I'll add a fixture.

**Comment by Fluorax** [comment]
Tested on master (`d9860eb`), gfx1100 (RX 7900 XTX), ROCm 6.2, Docker, `qwen3.6:27b.mq4`.

Two prompts tested, both broken, different failure modes:

**Prompt 1:** "what files are in this directory?"
```
<tool_call>
<tool_call>
... (×9)
{"arguments": {"param": {"path": "/home"}}
</think>
```
Empty stacked tags, no name field, malformed args, `</think>` leak.

**Prompt 2:** "write a file here named test.md with the text test inside"
```
<tool_call>
<tool_call>
... (×100+)
</think>
<tool_call>
... (continues)
```
Pure `<tool_call>` attractor loop, same mechanism as the section header repetition bug. Hits token limit without producing any JSON.

The fix in `parseToolCalls` doesn't get a chance to run on prompt 2, there's no JSON to repair, just the tag token in an infinite loop. This looks like a model-level attractor on the `<tool_call>` token itself under MQ4, not a parsing issue.

The `</think>` appearing mid-stream in both outputs suggests thinking mode is also interfering with tool call token generation.

**Comment by Kaden-Schutt** [comment]
@Fluorax — both repro queries you posted are now clean on master after PR #121 lands.

Re-ran your exact prompts against `qwen3.6:27b.mq4` on a 7900 XTX (gfx1100, ROCm 7.2) with the new build:

```
Prompt 1: "what files are in this directory?"
→ tool_calls: [{ name: "bash", arguments: { command: "ls -la" } }]
  finish_reason: tool_calls

Prompt 2: "write a file here named test.md with the text test inside"
→ tool_calls: [{ name: "write", arguments: { path: "test.md", content: "test" } }]
  finish_reason: tool_calls
```

No `<tool_call>` attractor loop, no nested openers in the visible stream, no parser repair signal on stderr. The fix is two layers:

1. **Engine-side (daemon)** — new GPU-side block that suppresses the `<tool_call>` token whenever the recent decoded window has 2+ unclosed openers (`opens − closes >= 2`). This trips before the model can stack a third nested opener. Same for `<think>` to head off the boundary corruption you saw. Per-step cost: 0 µs when not tripped, ~5 µs when tripped (single 4-byte H2D into the logits buffer).

2. **CLI-side (parser)** — `parseToolCalls` now strips any leading `<tool_call>\s*` repeats from a captured block before JSON parse. Defense-in-depth in case any nested opener slips through (it shouldn't — the engine block fires first — but if calibration drift gets worse on a future model, the parser still recovers).

Both layers are tested:
- 10 Rust unit tests in `llama::tests` cover threshold edges, window scope, complete-pair-allow, depth-saturate-at-zero
- 14 Bun tests in `cli/parse_tool_calls.test.ts` (4 new for nested-opener strip including empty-body fall-through)
- Coherence-gate green on 6/6 prompts including the existing tool-call coverage test

This is still a stopgap on the symptom — the underlying root cause is MQ4 calibration drift on structured-output token positions. Path C calibration retrain (#39, custom DFlash draft retrain on MI300X) is the proper fix and we'll come back to it. But for the practical "agent harness can't dispatch tool calls on MQ4" problem, master should now work for you.

Will close once PR #121 ships; please re-test on master and confirm if you find new failure shapes.

**Comment by Kaden-Schutt** [comment]
The fix is in **v0.1.9-alpha.1** (released just now, marked latest):
https://github.com/Kaden-Schutt/hipfire/releases/tag/v0.1.9-alpha.1

Linux: `hipfire update` (or `git pull && cargo install --path crates/engine`).
Windows: re-run `install.ps1` — the dynamic-release-query will pull the fresh `daemon.exe` + `infer.exe` + `run.exe` automatically.

Leaving this open until you confirm your harness is dispatching tool calls correctly on your end. If you hit any new failure shape, paste the raw `<tool_call>` block and I'll add it to the test suite.

**Comment by Fluorax** [comment]
Confirmed fixed on master 80b4d065b5b6c23ee18f8433a266aa589156ae78. Attractor loop on <tool_call> token is gone, both test prompts dispatch and execute correctly.
Two remaining issues:

1) USER literal: model uses USER instead of actual username on first attempt. Harness recovered via echo $USER and retried successfully. Not blocking.
2) Response summarization loop: after the tool executes and returns results, the model loops on the output content (e.g. .android.jsonc repeated ~10 times). This is the general structured output attractor, separate from the tool call fix. Still present.

Tool call dispatch is fixed. The content generation attractor is a separate issue.


---



**Edit 2:  follow-up testing on master `80b4d065`:**

The pure stacked opener loop is gone, that part is fixed. Found three new patterns:

**Pattern 1  `</think>` interleaved attractor (bypasses depth block):**
```
<tool_call>
<|im_start|>assistant
</think>
<tool_call>
</think>
</think>
I can see from the previous listing that ~ contains a scripts directory. Let me check what's in ~/scripts:

<tool_call>
</think>
</think>
... (repeats until token limit)
```
The `opens − closes >= 2` block doesn't fire because `</think>` tokens interspersed between openers reset the depth counter. Also seeing `<|im_start|>assistant` leaking into the output stream.

**Pattern 2  `param` key drift:**
```json
{"param": "ls -la ~/scripts"}
```
Should be `{"command": "ls -la ~/scripts"}`. Validation fails. On retry the model switches to the wrong tool entirely.

**Pattern 3  broken JSON opening:**
After explicit correction ("use command instead of param"), model produces correct key but malformed JSON:
```
{"
   "command": "ls -la ~/"
}
```
Orphaned quote after `{` on first line. Not valid JSON — parser can't extract args, call is dropped silently.

All three share the same root cause: MQ4 calibration drift on structured token positions.

---

**Comment by Kaden-Schutt** [comment]
Fluorax confirmed fixed on master `80b4d06` after PR #121 (`3f39668` GPU-side attractor block + `e932811` defensive parseToolCalls repair). Shipped in v0.1.9-alpha.1.

Two follow-ups from the confirmation comment (USER literal substitution + occasional repetition with default sampler) are separable and tracked under the broader Qwen3.6 27B tool-call agenting surface (`feedback/qwen36_27b_dflash_draft_misaligned_2026_04_26` + #87). Reopen only if the original `<plain>...</param>` malformation returns.
